### PR TITLE
 Fix overlapping numbers for stacked bar renderer

### LINKED ISSUE #319: Stacked bar renderer numbers overlap when hovering the bars
* Release number or git hash: 
 * Web browser version and OS: Chrome  81.0.4044.129 / Ubuntu

### Steps to reproduce

 1. Add a weighted sum column
 1. Choose the renderer  to be the stacked bar renderer

### Observed behavior
 * For very low values the rows' values show on top of each other
### Expected behavior
 * The column values should not overlap

### PR DESCRIPTION
**Pull request by sgratzl**
closes #319 

**prerequisites**: 
 * [x] branch is up-to-date with the branch to be merged with, i.e. develop
 * [x] build is successful
 * [x] code is cleaned up and formatted 
 * [ ] tested with Firefox 52, Firefox 57+, Chrome 64+, MS Edge 16+

### Summary
 
image

image

### CODE REVIEW

**Comment by sgratzl** [comment]
> Could we make this more consistent (i.e., avoid the flickering and that the color switches at a certain threshold but then remains like this)?

atm. the larger bar determines the adaptive foreground color. I'm not sure how easy it is that for N bars, the text turns only to white for bars that need it. The logic for the one base case might be extendible for N cases

**Comment by sgratzl** [comment]
image

File: src/renderer/MultiLevelCellRenderer.ts [hunk 1]
```diff
@@ -125,6 +150,26 @@ export default class MultiLevelCellRenderer extends AAggregatedGroupRenderer<IMu
...
+        if (!overlapData) {
+          n.classList.remove(cssClass('stack-label'));
+          if (toWait.length > 0) {
+            return <IAbortAblePromise<void>>abortAbleAll(toWait);
+          }
+          return null;
+        }
+        n.classList.add(cssClass('stack-label'));
+
+        const combinedLabel = overlapData.map((col) => col.label).join('	');
```
Comment (thinkh) [inline src/renderer/MultiLevelCellRenderer.ts:163]:
```suggestion
        const combinedLabel = overlapData.map((col) => col.label).join(' '); // use "em space" (U+2003 &#8195)
```

Copied from http://emptycharacter.com/

**Comment by sgratzl** [comment]
> Maybe we can even align the hover values with the column header?

as you want, an open question is whether it should do that regardless of the bar size or just when the bar length is not wide enough?

**Comment by thinkh** [comment]
Your last commit improves the readability even more. 👍 

grafik
grafik


> > Maybe we can even align the hover values with the column header?
> 
> as you want, an open question is whether it should do that regardless of the bar size or just when the bar length is not wide enough?

Good question! I think both options, text inside the bar if there is enough space and text as grid above the stacked bar, might be useful.

I would like to test both options, if possible, for multiple rows and datasets. Would it be difficult to implement the second option (and provide a switch)?

**Comment by sgratzl** [comment]
see sgratzl/stacklabel2 branch

image

**Comment by thinkh** [comment]
Thanks for implementing the alternative. I tested both options and had an internal discussion about both options, with the result that they are both not very appealing.

We have two alternative suggestions:

1. Show the value for the nested columns, only if the bar is long enough. That would resolve the overlap or zebra effect. The result would look similar to Microsoft PowerBI stacked bar.
   grafik
2. Alternatively, we can display only the sum of all nested values at right aligned to the bar or the weighted sum column. In case the bar is 90-100% long the text will be superimposed as it is now. Individual values are only available on hover (see below).
    image

In both cases, since we are hiding some values, we would need a tooltip displaying the values of all nested columns.

@sgratzl What is your opinion? Which of the options above is easier to implement for you?

**Comment by sgratzl** [comment]
>     1. Show the value for the nested columns, only if the bar is long enough. That would resolve the overlap or zebra effect. The result would look similar to Microsoft PowerBI stacked bar.

so instead of a combined label, hide all bar labels where the label length is < bar length. 

Does the same apply to normal bar column and the last column in a stacked bar? 

>     2\. Alternatively, we can display only the sum of all nested values at right aligned to the bar or the weighted sum column. In case the bar is 90-100% long the text will be superimposed as it is now. Individual values are only available on hover (see below).

the sum value has no meaning to the user. since it is the weighted sum of normalized values.

**Comment by thinkh** [comment]
> so instead of a combined label, hide all bar labels where the label length is < bar length.

Yes, if you refer to the implementation.

> Does the same apply to normal bar column and the last column in a stacked bar?

I'm not sure if we want this for the normal bar column, too. We were discussing a similar display as in MS PowerBI. There the values are right aligned, which aligns with the convention to display numerical values and would minimize the zebra effect in our case.

grafik


> the sum value has no meaning to the user. since it is the weighted sum of normalized values.

I do not understand this point. The hover text always shows the column values and not the weighted values, right? So if `A = 5` and `B = 3` then the sum value behind the bar would show `8`.

**Comment by sgratzl** [comment]
> I do not understand this point. The hover text always shows the column values and not the weighted values, right? So if `A = 5` and `B = 3` then the sum value behind the bar would show `8`.

the hover values show the raw values, that is true. However, the sum is the weighted sum of normalized values after the mapping. Thus, in your example

```
A = 5, B = 3,
A domain: 0 ... 10
B domain: 0 ... 30
A_normalized = 0.5
B_normalized = 0.1
A_weight = 0.5
B_weight = 0.5
sum = 0.5 * 0.5 + 0.5 * 0.1 = 0.3
```

**Comment by thinkh** [comment]
> the hover values show the raw values, that is true. However, the sum is the weighted sum of normalized values after the mapping. Thus, in your example
> 
> ```
> A = 5, B = 3,
> A domain: 0 ... 10
> B domain: 0 ... 30
> A_normalized = 0.5
> B_normalized = 0.1
> A_weight = 0.5
> B_weight = 0.5
> sum = 0.5 * 0.5 + 0.5 * 0.1 = 0.3
> ```

Thanks for the explanation. I agree that displaying the `0.3` does not mean much to the user. Maybe we can display the sum of the weighted raw values instead as it reflects what the user sees in the total bar length. We could explain this sum value in the tooltip as _weighted sum_ along with the other raw values.

```
A = 5, B = 3,
A_weight = 0.5
B_weight = 0.5
sum = 0.5 * 5 + 0.5 * 3 = 4
```

If the whole idea of showing the summed value is not possible we can pursue the alternative

**Comment by sgratzl** [comment]
> If the whole idea of showing the summed value is not possible

in your example you ignored the different domains which would lead to:
```
A domain: 0 ... 10
B domain: 0 ... 30

row1:
A = 5, B = 3,
A_normalized = 0.5
B_normalized = 0.1
A_weight = 0.5
B_weight = 0.5
normalized_sum = 0.5 * 0.5 + 0.5 * 0.1 = 0.3
sum = 0.5 * 5 + 0.5 * 3 = 4

row2:
A = 2, B = 9
A_normalized = 0.2
B_normalized = 0.3
normalized_sum = 0.5 * 0.2 + 0.5 * 0.3 = 0.25
sum = 0.5 * 2 + 0.5 * 9 = 5.5

```

in case of sorting the values descending row 1 would be before row 2 even tho their labels indicate otherwise.

**Comment by sgratzl** [comment]
hide labels if not enough space:

image

**Comment by sgratzl** [comment]
should this be part of v4.0.0, ?

**Comment by sgratzl** [comment]
> @sgratzl Could you please change this one more time?

see https://github.com/lineupjs/lineupjs/pull/337 since I can undo most of the changes

**Comment by sgratzl** [comment]
> @sgratzl What do you think? Do you have a different idea?

one could add a custom toolbar icon which essentially just toggles between the two renderers

**Comment by thinkh** [comment]
> see #337 since I can undo most of the changes

Thanks for giving this another try. I will merge that PR #337 and close this PR.


> one could add a custom toolbar icon which essentially just toggles between the two renderers
> 
> as in [codepen.io/sgratzl/pen/wvMKvxZ](https://codepen.io/sgratzl/pen/wvMKvxZ)

Nice solution! Is that something you can add it to the library or we should add to our applications?

**Comment by sgratzl** [comment]
> Nice solution! Is that something you can add it to the library or we should add to our applications?

for now I would suggest not adding it to the library, since it uses some assumptions about renderers which cannot be fulfilled in a general way

**Comment by thinkh** [comment]
> for now I would suggest not adding it to the library, since it uses some assumptions about renderers which cannot be fulfilled in a general way

Ok, that's fine for now.

**Pull request closed without merging**